rllib/examples/learners/classes/mixture_of_gaussian_learner.py

```diff
@@ -0,0 +1,219 @@
+                The predicted mixture weights of the MoG for the current timestep.
+
+        Returns:
+            summing_log : torch.Tensor
+                The computed negative log-likelihood loss for the given inputs.
+        """
+
+        td_targets_expanded = td_targets.unsqueeze(1)
+        sigma_clamped = sigma_current
+        log_2_pi = torch.log(2 * torch.tensor(math.pi))
+        factor = -torch.log(sigma_clamped) - 0.5 * log_2_pi
+        mus = td_targets_expanded - mu_current
+
+        logp = torch.clamp(
+            factor - torch.square(mus) / (2 * torch.square(sigma_clamped)), -1e10, 10
+        )
+        # Little trick of using log_softmax on the current alphas to prevent nans
+        loga = torch.clamp(nn.functional.log_softmax(alpha_current, dim=-1), 1e-6, None)
+
+        summing_log = -torch.logsumexp(logp + loga, dim=-1)
+        return summing_log
+
+    @override(PPOTorchLearner)
+    def compute_loss_for_module(
+        self,
+        *,
+        module_id: ModuleID,
+        config: PPOConfig,
+        batch: Dict[str, Any],
+        fwd_out: Dict[str, TensorType],
+    ) -> TensorType:
+
+        if Columns.LOSS_MASK in batch:
+            mask = batch[Columns.LOSS_MASK]
+            num_valid = torch.sum(mask)
+
+            def possibly_masked_mean(data_):
+                return torch.sum(data_[mask]) / num_valid
+
+        else:
+            possibly_masked_mean = torch.mean
+
+        module = self.module[module_id].unwrapped()
+
+        action_dist_class_train = module.get_train_action_dist_cls()
+        action_dist_class_exploration = module.get_exploration_action_dist_cls()
+
+        curr_action_dist = action_dist_class_train.from_logits(
+            fwd_out[Columns.ACTION_DIST_INPUTS]
+        )
+        prev_action_dist = action_dist_class_exploration.from_logits(
+            batch[Columns.ACTION_DIST_INPUTS]
+        )
+
+        logp_ratio = torch.exp(
+            curr_action_dist.logp(batch[Columns.ACTIONS]) - batch[Columns.ACTION_LOGP]
+        )
+
+        # Only calculate kl loss if necessary (kl-coeff > 0.0).
+        if config.use_kl_loss:
+            action_kl = prev_action_dist.kl(curr_action_dist)
+            mean_kl_loss = possibly_masked_mean(action_kl)
+        else:
+            mean_kl_loss = torch.tensor(0.0, device=logp_ratio.device)
+
+        curr_entropy = curr_action_dist.entropy()
+        mean_entropy = possibly_masked_mean(curr_entropy)
+
+        surrogate_loss = torch.min(
+            batch[Postprocessing.ADVANTAGES] * logp_ratio,
+            batch[Postprocessing.ADVANTAGES]
+            * torch.clamp(logp_ratio, 1 - config.clip_param, 1 + config.clip_param),
+        )
+
+        # Compute the MOG value loss using the negative log-likelihood
+        if config.use_critic:
+            rewards = batch["rewards"]
+            dones = batch["dones"]
+            gamma = config["gamma"]
+
+            mog_components = batch[Columns.MOG_COMPONENTS]
```

Comment:
`Columns`' constants are reserved for often used constants. Custom constants used in learners/modules go into the custom learner modules. I can change this, no problem.